File: problems/2026/04/2026-04-12-leetcode-146-LRU缓存/SolutionLC146.java
import java.util.HashMap;
import java.util.Map;

/**
 * LeetCode 146：LRU 缓存（哈希表 + 双向链表，get/put 均 O(1)）
 * <p>
 * 提交 LeetCode 时可将内部类 {@link LRUCache} 单独复制为顶层类使用。
 */
public class SolutionLC146 {

    static class LRUCache {

        private static class Node {
            int key;
            int val;
            Node prev;
            Node next;

            Node(int key, int val) {
                this.key = key;
                this.val = val;
            }
        }

        private final int capacity;
        private final Map<Integer, Node> map = new HashMap<>();
        private final Node head = new Node(0, 0);
        private final Node tail = new Node(0, 0);

        public LRUCache(int capacity) {
            this.capacity = capacity;
            head.next = tail;
            tail.prev = head;
        }

        public int get(int key) {
            Node node = map.get(key);
            if (node == null) {
                return -1;
            }
            remove(node);
            addToTail(node);
            return node.val;
        }

        public void put(int key, int value) {
            Node existing = map.get(key);
            if (existing != null) {
                existing.val = value;
                remove(existing);
                addToTail(existing);
                return;
            }
            if (map.size() == capacity) {
                Node lru = head.next;
                remove(lru);
                map.remove(lru.key);
            }
            Node node = new Node(key, value);
            map.put(key, node);
            addToTail(node);
        }

        private void remove(Node node) {
            node.prev.next = node.next;
            node.next.prev = node.prev;
        }

        /** 接到尾哑元前，表示最近使用 */
        private void addToTail(Node node) {
            node.prev = tail.prev;
            node.next = tail;
            tail.prev.next = node;
            tail.prev = node;
        }
    }

    public static void main(String[] args) {
        LRUCache cache = new LRUCache(2);
        cache.put(1, 1);
        cache.put(2, 2);
        System.out.println("get(1) 预期 1: " + cache.get(1));
        cache.put(3, 3);
        System.out.println("get(2) 预期 -1: " + cache.get(2));
        cache.put(4, 4);
        System.out.println("get(1) 预期 -1: " + cache.get(1));
        System.out.println("get(3) 预期 3: " + cache.get(3));
        System.out.println("get(4) 预期 4: " + cache.get(4));
    }
}

File: problems/2026/04/2026-04-12-leetcode-146-LRU缓存/SolutionLC146.py
from typing import Dict, Optional


class _Node:
    __slots__ = ("key", "val", "prev", "next")

    def __init__(self, key: int = 0, val: int = 0):
        self.key = key
        self.val = val
        self.prev: Optional["_Node"] = None
        self.next: Optional["_Node"] = None


class LRUCache:
    """LeetCode 146：哈希表 + 双向链表，O(1) get / put"""

    def __init__(self, capacity: int):
        self._cap = capacity
        self._map: Dict[int, _Node] = {}
        self._head = _Node()
        self._tail = _Node()
        self._head.next = self._tail
        self._tail.prev = self._head

    def get(self, key: int) -> int:
        if key not in self._map:
            return -1
        node = self._map[key]
        self._remove(node)
        self._add_to_tail(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self._map:
            node = self._map[key]
            node.val = value
            self._remove(node)
            self._add_to_tail(node)
            return
        if len(self._map) == self._cap:
            lru = self._head.next
            assert lru is not None
            self._remove(lru)
            del self._map[lru.key]
        node = _Node(key, value)
        self._map[key] = node
        self._add_to_tail(node)

    def _remove(self, node: _Node) -> None:
        assert node.prev is not None and node.next is not None
        node.prev.next = node.next
        node.next.prev = node.prev

    def _add_to_tail(self, node: _Node) -> None:
        assert self._tail.prev is not None
        p = self._tail.prev
        p.next = node
        node.prev = p
        node.next = self._tail
        self._tail.prev = node


if __name__ == "__main__":
    c = LRUCache(2)
    c.put(1, 1)
    c.put(2, 2)
    print("get(1):", c.get(1))
    c.put(3, 3)
    print("get(2):", c.get(2))
    c.put(4, 4)
    print("get(1):", c.get(1))
    print("get(3):", c.get(3))
    print("get(4):", c.get(4))
